core/client/fs/src/main/java/alluxio/client/file/FileSystem.java

```diff
@@ -148,10 +149,10 @@ private static FileSystem create(FileSystemContext context, boolean cachingEnabl
           LOG.debug("{}={} ({})", key.getName(), value, source);
         }
       }
-      if (context.getClusterConf().getBoolean(PropertyKey.USER_METADATA_CACHE_ENABLED)) {
-        return new CachingFileSystem(context, cachingEnabled);
-      }
-      return BaseFileSystem.create(context, cachingEnabled);
+      Class fsClass = context.getClusterConf().getClass(PropertyKey.USER_FILESYSTEM_CLASS);
+      Class[] ctorArgClasses = new Class[] {FileSystemContext.class, boolean.class};
+      Object[] ctorArgs = new Object[] {context, cachingEnabled};
+      return (FileSystem) CommonUtils.createNewClassInstance(fsClass, ctorArgClasses, ctorArgs);
```

Comment:
I'm not a huge fan of creating the classes this way because it creates an implicit rather than explicit contract on all implementations to use the exact same constructor. I don't think it's the best approach.

Because we don't have many FS implementations I think I'm okay leaving it for now, but we might want to consider a factory-based pattern or something similar if we end up extending this more in the future.

Reply:
I agree, even in our current code we don't have a standard way of creating the file system